Break up arguments section (#2586)
pallets/click: docs/arguments.rst

    invoke(touch, args=['foo.txt'])

An argument may be assigned a :ref:`parameter type <parameter-types>`. If no type is provided, the type of the default value is used. If no default value is provided, the type is assumed to be STRING.

.. admonition:: Note on Required Arguments

>    It is possible to make an argument required by setting ``required=True``.  It is not recommended since we think command line tools should gracefully degrade into becoming no ops.  We think this because command line tools are often invoked with wildcard inputs and they should not error out if the wildcard is empty.

Author: If you read the full docs, their are some opinionated pieces that look to be written by the original author and correspond to unix command line tool practices. This is one of them so I modified it slightly and left it in place. I was planning on leaving them in place on my first pass and then we could make a decision as a whole on how opinionated we wanted to be.

>    It is possible to make an argument required by setting ``required=True``.  It is not recommended since we think command line tools should gracefully degrade into becoming no ops.  We think this because command line tools are often invoked with wildcard inputs and they should not error out if the wildcard is empty.

Author: If you ignore the heading on the note which is the almost the same as the paragraph it was in, then it seems the author was talking about required arguments in general.

>    It is possible to make an argument required by setting ``required=True``.  It is not recommended since we think command line tools should gracefully degrade into becoming no ops.  We think this because command line tools are often invoked with wildcard inputs and they should not error out if the wildcard is empty.

Reviewer: Makes sense. Could we rephrase "into becoming no ops" here? I think what is meant is that only using arguments makes a cli look like ancient magics, right?

Could we still document `nargs=-1` behaviour with `required=True` however? I agree that only using arguments is bad. Though there's a definite use case for having a variadic argument that requires at least one.

>    It is possible to make an argument required by setting ``required=True``.  It is not recommended since we think command line tools should gracefully degrade into becoming no ops.  We think this because command line tools are often invoked with wildcard inputs and they should not error out if the wildcard is empty.

Author: I think it means that if you invoke the tool with no arguments then it should do nothing rather than failing with an error. I am going to leave in alone for now and open a meta issue on opinionated takes about things.

> Multiple Arguments

Author: Variadic refers to taking an indefinite number of arguments. ie the special case of nargs = -1

In mathematics and in computer programming, a variadic function is a function of indefinite arity, i.e., one which accepts a variable number of arguments

https://en.wikipedia.org/wiki/Variadic_function 

So I don't think its the right name for the heading. But I see how this is confusing, multiple arguments is used to mean something different in the options code. I will look for a different word than multiple

> Multiple Arguments

Author: I think those are more literally true but at the cost of being very long. I am going to leave this as Multiple Arguments for now. Feel free to add more examples if you like.

-----------------------------------

> To set the number of argument use the ``nargs`` kwarg. It can be set to any positive integer and -1. Setting it to -1, makes the number of arguments arbitrary (which is called variadic) and can only be used once. The arguments are then packed as a tuple and passed to the function.

Reviewer: I think the previous explanation was more correct and clear. This confuses "multiple arguments" and "variadic arguments". See my previous comment.

Author: In my view variadic is a special case of taking more than one argument, but that may not be the clearest way for documentation purposes. See above.

.. click:example::

    @click.command()
    @click.argument('src', nargs=1)
    @click.argument('dsts', nargs=-1)

>     def copy(src, dsts):

Reviewer: ```suggestion
    def copy(src: str, dsts: tuple[str, ...]):
```

        """Move file SRC to DST."""
        for destination in dsts:
            click.echo(f"Copy {src} to folder {destination}")

And what it looks like:

.. click:run::

    invoke(copy, args=['foo.txt', 'usr/david/foo.txt', 'usr/mitsuko/foo.txt'])

> .. admonition:: Note on Handling Files.

Reviewer: ```suggestion
.. admonition:: Note on Handling Files
```

>     This is not how you would handle files and files paths, since it passes them in as strings. For more see :ref:`handling-files`.

Reviewer: ```suggestion
    This is not how you should handle files and files paths. This merely used as a simple example. See :ref:`handling-files` to learn more about how to handle files in parameters.
```

Argument Escape Sequences
---------------------------

If you want to process arguments that look like options, like a file named ``-foo.txt`` or ``--foo.txt`` , you must pass the ``--`` separator first. After you pass the ``--``, you may only pass arguments. This is a common feature for POSIX command line tools.

Example usage:

.. click:example::

    @click.command()

>     @click.argument('files', nargs=-1, type=click.Path())
>     def touch(files):

Reviewer: ```suggestion
    @click.argument('files', nargs=-1, type=click.Path(path_type=pathlib.Path))
    def touch(files: tuple[pathlib.Path, ...]):
```

I feel like we should make `path_type=pathlib.Path` the default. Until it's not, let's maybe make it a bit more clear they can get a `Path` type here as it otherwise defaults to `str`.

        """Print all FILES file names."""
        for filename in files:
            click.echo(filename)

And from the command line:

.. click:run::

    invoke(touch, ['--', '-foo.txt', 'bar.txt'])

If you don't like the ``--`` marker, you can set ignore_unknown_options to True to avoid checking unknown options:

.. click:example::

    @click.command(context_settings={"ignore_unknown_options": True})
    @click.argument('files', nargs=-1, type=click.Path())
    def touch(files):
        """Print all FILES file names."""
        for filename in files:
            click.echo(filename)

And from the command line:

.. click:run::

    invoke(touch, ['-foo.txt', 'bar.txt'])

.. _environment-variables:

Environment Variables
---------------------

> Arguments can use environment variables. To do so pass the name of the environment variable(s) into ``click.argument``.

Reviewer: ```suggestion
Arguments can use environment variables. To do so, pass the name(s) of the environment variable(s) via `envvar` in ``click.argument``.
```

Argument Escape Sequences
---------------------------

If you want to process arguments that look like options, like a file named ``-foo.txt``, you must pass the ``--`` separator first. After you pass the ``--``, you may only pass arguments. This is a common feature for POSIX command line tools.

Example usage:

.. click:example::

    @click.command()
    @click.argument('files', nargs=-1, type=click.Path())
    def touch(files):
        """Print all FILES file names."""
        for filename in files:
            click.echo(filename)

And from the command line:

.. click:run::

    invoke(touch, ['--', '-foo.txt', 'bar.txt'])

If you don't like the ``--`` marker, you can set ignore_unknown_options to True to avoid checking unknown options:

.. click:example::

    @click.command(context_settings={"ignore_unknown_options": True})
    @click.argument('files', nargs=-1, type=click.Path())
    def touch(files):
        """Print all FILES file names."""
        for filename in files:
            click.echo(filename)

And from the command line:

.. click:run::

    invoke(touch, ['-foo.txt', 'bar.txt'])

.. _environment-variables:

Environment Variables
---------------------

> This feature is not recommended since it be confusing to users. Arguments can only pull environment variables from ? explicitly named environment variables. In that case, it can also be a list of different environment variables where the first one is picked. ?

Author: @AndreasBackx Do you know what the documentation could mean in this section? I have not a clue so providing clearer wording and a better example is difficult.

Author: @AndreasBackx I think actually only Armin knew. It was basically unchanged since the first commit. But I figured it out by playing around. Could you give this PR a once over?
